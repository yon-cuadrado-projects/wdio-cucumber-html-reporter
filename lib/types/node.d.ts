import { Browser } from 'webdriverio';
import { cjson_metadata } from '../models';
export { };

declare global {
    namespace NodeJS {
        export interface EnvironmentVariables {
            wdioCucumberHtmlReporter: string;
            attachment: string;
        }

        export interface Global {
            browser: Browser<'sync'>;
        }
    }

    namespace WebDriver {
        export interface W3CCapabilities {
            cjson_metadata: cjson_metadata;
        }
    }
}
