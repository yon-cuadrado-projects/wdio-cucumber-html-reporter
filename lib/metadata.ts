import { AppData, BrowserData, cjson_metadata } from './models';
import { DesiredCapabilitiesExtended, RunnerStatsExtended, W3CCapabilitiesExtended, WebdriverIOExtended } from './types/wdio';
import { Models } from 'cucumber-html-report-generator';
import { NOT_KNOWN } from './constants';
import WebDriver from 'webdriver';

export class Metadata {
    /**
     * ```
     * Determine the metadata that needs to be added
     *
     * @param {object} data instance data
     *
     * @returns {
     *  {
     *      metadata: {
     *          app: {
     *              name: string,
     *              version: string
     *          },
     *          browser: {
     *              name: string,
     *              version: string
     *          },
     *          device: string,
     *          platform: {
     *              name: string,
     *              version: string
     *          }
     *      }
     *  }
     * }
     * ```
     */
    public determineMetadata ( data: RunnerStatsExtended ): Models.Metadata[] {
        const featureMetadata = <Models.Metadata[]>[];

        let instanceData: AppData | BrowserData;
        const currentCapabilities = data.capabilities as W3CCapabilitiesExtended;
        const optsCaps = browser.options.capabilities;
        const currentConfigCapabilities = data.capabilities as DesiredCapabilitiesExtended;
        const w3cCaps = ( browser.options as WebdriverIOExtended ).requestedCapabilities;
        const metadata: cjson_metadata = ( currentConfigCapabilities as W3CCapabilitiesExtended )['cjson:metadata'] // For WDIO V6
            || w3cCaps?.['cjson:metadata'] // When an app is used to test
            || ( optsCaps as DesiredCapabilitiesExtended )?.['cjson:metadata'] // devtools
            || {} as cjson_metadata;

        // When an app is used to test
        // eslint-disable-next-line @typescript-eslint/tslint/config
        if ( currentConfigCapabilities.app || ( currentConfigCapabilities ).testobject_app_id || metadata.app ) {
            instanceData = this.determineAppData( currentConfigCapabilities, metadata );
            featureMetadata.push( <Models.Metadata>{ name: 'application',value: `${ instanceData.app.name} ${instanceData.app.version}` } );
        } else {
            // Then a browser
            instanceData = this.determineBrowserData( currentCapabilities, currentConfigCapabilities, metadata );
            featureMetadata.push( <Models.Metadata>{ name: 'browser',value: `${ instanceData.browser.name} ${instanceData.browser.version}` } );
        }

        const device = this.determineDeviceName( metadata, currentConfigCapabilities );
        const platform = {
            name: this.determinePlatformName( metadata, currentCapabilities ),
            version: this.determinePlatformVersion( metadata ),
        };

        featureMetadata.push( <Models.Metadata>{ name: 'device', value: device } );
        featureMetadata.push( <Models.Metadata>{ name: 'platform', value: `${platform.name} ${platform.version}` } );

        return featureMetadata;
    }

    /**
     * Determine the device name
     *
     * @param {object} metadata
     * @param {object} currentConfigCapabilities
     * @return {string}
     */
    public determineDeviceName ( metadata: cjson_metadata, currentConfigCapabilities: WebDriver.DesiredCapabilities ): string {
        return ( metadata.device || currentConfigCapabilities.deviceName || `Device name ${NOT_KNOWN}` );
    }

    /**
     * Determine the platform name
     *
     * @param {object} metadata
     * @param {object} currentCapabilities
     * @return {string}
     */
    public determinePlatformName ( metadata: cjson_metadata, currentCapabilities: WebDriver.DesiredCapabilities ): string {
        const currentPlatformName = currentCapabilities.platformName
            ? currentCapabilities.platformName.includes( 'mac' )
                ? 'osx'
                : currentCapabilities.platformName.includes( 'windows' )
                    ? 'windows'
                    : currentCapabilities.platformName
            : `Platform name ${NOT_KNOWN}`;
        return ( metadata.platform && metadata.platform.name )
            ? metadata.platform.name
            : currentPlatformName;
    }

    /**
     * Determine the platform version
     *
     * @param {object} metadata
     * @return {string}
     */
    public determinePlatformVersion ( metadata: cjson_metadata ): string {
        return ( metadata && metadata.platform && metadata.platform.version )
            ? metadata.platform.version
            : `Version ${NOT_KNOWN}`;
    }

    /**
     * Determine the app data
     *
     * @param {object} currentConfigCapabilities The capabilities from the configured capabilities
     * @param {object} metadata The custom set capabilities
     *
     * @returns {
     * {
     * app: {
     *          name: string,
     *          version: string,
     *      },
     *  }
     * }
     */
    public determineAppData ( currentConfigCapabilities: DesiredCapabilitiesExtended, metadata: cjson_metadata ): AppData {
        const metaAppName: string = ( metadata.app && metadata.app.name ) ? metadata.app.name : 'No metadata.app.name available';
        const metaAppVersion: string = ( metadata.app && metadata.app.version ) ? metadata.app.version : 'No metadata.app.version available';
        const appPath = currentConfigCapabilities.app || currentConfigCapabilities.testobject_app_id || metaAppName;
        const appName = appPath.substring( appPath.replace( '\\', '/' ).lastIndexOf( '/' ) ).replace( '/', '' );

        return {
            app: {
                name: appName,
                version: metaAppVersion,
            },
        };
    }

    /**
     * Determine the browser data
     *
     * @param {object} currentCapabilities The capabilities of the current run, that holds the most actual data
     * @param {object} currentConfigCapabilities The capabilities from the configured capabilities
     * @param {object} metadata The custom set capabilities
     *
     * @returns {
     *  {
     *      browser: {
     *          name: string,
     *          version: string,
     *      },
     *  }
     * }
     */
    public determineBrowserData ( currentCapabilities: WebDriver.DesiredCapabilities, currentConfigCapabilities: WebDriver.DesiredCapabilities, metadata: cjson_metadata ): BrowserData {
        const browserName = currentCapabilities.browserName
            || currentConfigCapabilities.browserName
            || ( ( metadata && metadata.browser && metadata.browser.name ) ? metadata.browser.name : 'No metadata.browser.name available' );
        const browserVersion = currentCapabilities.version
            || currentCapabilities.browserVersion
            || currentConfigCapabilities.browserVersion
            || ( ( metadata && metadata.browser && metadata.browser.version ) ? metadata.browser.version : 'No metadata.browser.version available' );

        return <BrowserData>{
            browser: {
                name: browserName,
                version: browserVersion,
            }
        };
    }
}
