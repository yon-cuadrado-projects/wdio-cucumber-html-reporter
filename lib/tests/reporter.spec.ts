import * as fs from 'fs';
import { DEFAULT_LANGUAGE, FAILED, PASSED, PENDING, TEXT_PLAIN } from '../constants';
import { HookStatsExtended, RunnerStatsExtended, SuiteStatsExtended, TestStatsExtended } from '../types/wdio';
import { Report, Scenario } from '../models';
import { copySync, readJsonSync, readdirSync, removeSync } from 'fs-extra';
import CucumberHtmlJsonReporter from '../reporter';
import { FULL_RUNNER_STATS } from './__mocks__/mocks';
import { Metadata } from '../metadata';
import type { Models } from 'cucumber-html-report-generator';
import { fileExists } from './fileExists';
import path from 'path';

describe( 'reporter', () => {
    let tmpReporter: CucumberHtmlJsonReporter = null;
    const logFolder = '.tmp';
    const logFileName = 'logFile.json';
    const logFolderPath = path.join( __dirname, '../../', logFolder );
    const logFilePath = path.join( logFolderPath, logFileName );
    const language = DEFAULT_LANGUAGE;
    beforeAll( () => {
        if ( !fs.existsSync( logFolderPath ) ) {
            fs.mkdirSync( logFolderPath );
            fs.closeSync( fs.openSync( logFilePath, 'w' ) );
        }
    } );

    beforeEach( () => {
        tmpReporter = new CucumberHtmlJsonReporter( <Models.ReportGeneration>{}, language );
    } );

    describe( 'on create', () => {
        it( 'should set the defaults only if the reportProperties option is provided', () => {
            const noOptionsReporter = new CucumberHtmlJsonReporter( <Models.ReportGeneration>{ jsonDir: './tmp' } );
            expect( noOptionsReporter.options ).toMatchSnapshot();
        } );

        it( 'should verify initial properties', () => {
            expect( tmpReporter.options ).toMatchSnapshot();
            expect( tmpReporter.instanceMetadata ).toBeNull();
            tmpReporter.onRunnerStart( FULL_RUNNER_STATS );
            expect( tmpReporter.report ).toMatchSnapshot();
            tmpReporter.report = {
                'feature':{
                    'elements': [],
                } } ;

            expect( tmpReporter.report ).toMatchSnapshot();
        } );
    } );

    describe( 'onRunnerStart', () => {
        it( 'should set instance data if it is not available yet', () => {
            const metadata = <Models.Metadata[]>[{ name: 'browser', value: 'chrome 85' }];
            const metadataClassObject: Metadata = tmpReporter.metadataClassObject;
            const determineMetadataSpy: jest.SpyInstance = jest.spyOn( metadataClassObject, 'determineMetadata' ).mockReturnValue( metadata );

            expect( tmpReporter.instanceMetadata ).toBeNull();

            tmpReporter.onRunnerStart( {} as RunnerStatsExtended );

            expect( determineMetadataSpy ).toHaveBeenCalled();
            expect( tmpReporter.instanceMetadata ).toEqual( metadata );
        } );

        it( 'should set not set instance data if it is already available', () => {
            const metadata = <Models.Metadata[]>[{ name: 'browser', value: 'chrome 85' }];
            const determineMetadataSpy: jest.SpyInstance = jest.spyOn( new Metadata(), 'determineMetadata' ).mockReturnValue( metadata );

            tmpReporter.instanceMetadata = metadata;
            expect( tmpReporter.instanceMetadata ).toEqual( metadata );

            tmpReporter.onRunnerStart( {} as RunnerStatsExtended );

            expect( determineMetadataSpy ).not.toHaveBeenCalled();
        } );
    } );

    describe( 'onSuiteStart', () => {
        it( 'should add the CucumberJS feature object if it is not available', () => {
            tmpReporter = new CucumberHtmlJsonReporter( <Models.ReportGeneration>{}, language, logFilePath );
            expect( tmpReporter.report ).toMatchSnapshot();
            tmpReporter.report = <Report>{
                feature: {}
            };

            tmpReporter.onSuiteStart( { keyword: 'feature', title: 'test', uid: '' } as SuiteStatsExtended );

            expect( tmpReporter.report ).toMatchSnapshot();
        } );

        it( 'should add a scenario to the feature if the feature is already there', () => {
            tmpReporter = new CucumberHtmlJsonReporter( <Models.ReportGeneration>{}, language, logFilePath );

            // tmpReporter.report.feature = EMPTY_FEATURE;
            expect( tmpReporter.report.feature?.elements?.length ).toBe( undefined );
            tmpReporter.onSuiteStart( { keyword: 'feature', title: 'test', uid: '' } as SuiteStatsExtended );
            expect( tmpReporter.report.feature.elements ).toMatchSnapshot();
        } );
    } );

    describe( 'onHookStart', () => {
        it( 'should add a pending before step', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                    }]
                }
            };
            tmpReporter.onHookStart( { title: '', uid: '' } as HookStatsExtended );
            expect( tmpReporter.report.feature.elements[0].steps[0] ).toMatchSnapshot();
        } );

        it( 'should add a pending after step', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            result: {
                                status: PENDING
                            },
                            keyword: 'Given'
                        }]
                    }]
                }
            };
            tmpReporter.onHookStart( { title: '', uid: '', keyword: 'Given' } as HookStatsExtended );
            expect( tmpReporter.report.feature.elements[0].steps[1] ).toMatchSnapshot();
        } );
    } );

    describe( 'onHookEnd', () => {
        it( 'should update the last step of the current scenario with its result', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            keyword: 'Given'
                        }]
                    }]
                }
            };
            tmpReporter.onHookEnd( { title: '', uid: '' } as HookStatsExtended );

            expect( tmpReporter.report.feature.elements[0].steps[0].result.status ).toBe( PASSED );
        } );

        it( 'should call update a hook step to the current state when there is an error', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            keyword: 'Given'
                        }]
                    }]
                }
            };

            tmpReporter.onHookEnd( { state: FAILED, title: '', uid: '', error: new Error( 'error 1' ) } as HookStatsExtended );

            expect( tmpReporter.report.feature.elements[0].steps[0].result.status ).toBe( FAILED );
        } );
    } );

    describe( 'onTestStart', () => {
        it( 'should add the step data when a test is started', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: []
                    }]
                }
            };

            const argument = {
                'rows': [
                    {
                        'cells': [
                            'Cucumber',
                            'Cucumis sativus'
                        ]
                    },
                    {
                        'cells': [
                            'Burr Gherkin',
                            'Cucumis anguria'
                        ]
                    }
                ]
            };
            tmpReporter.onTestStart( { foo: 'bar', title: '', uid: '', argument, error: new Error( 'error 1' ) } as TestStatsExtended );

            expect( tmpReporter.report.feature.elements[0].steps[0].result.status ).toBe( '' );
        } );
    } );

    describe( 'onTestPass', () => {
        it( 'should update the last step with the provided data', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            result: {
                                status: PASSED
                            }
                        }]
                    }]
                }
            };
            tmpReporter.onTestPass( { state: PASSED, title: '', uid: '' } as TestStatsExtended );

            expect( tmpReporter.report.feature.elements[0].steps[0].result.status ).toBe( PASSED );
        } );
    } );

    describe( 'onTestFail', () => {
        it( 'should update the last step with the provided data', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            keyword: 'Given'
                        }]
                    }]
                }
            };

            tmpReporter.onTestFail( { state: FAILED, title: '', uid: '', error: new Error( 'error 1' ) } as TestStatsExtended );

            expect( tmpReporter.report.feature.elements[0].steps[0].result.status ).toBe( FAILED );
        } );
    } );

    describe( 'onTestSkip', () => {
        it( 'should call update a step', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            keyword: 'Given'
                        }]
                    }]
                }
            };

            tmpReporter.onTestSkip( { state: PENDING, title: '', uid: '' } as TestStatsExtended );

            expect( tmpReporter.report.feature.elements[0].steps[0].result.status ).toBe( PENDING );
        } );
    } );

    describe( 'onRunnerEnd', () => {
        it( 'should store the json file on the file system', async () => {
            const outputFolder = path.join( process.cwd(),'./.tmp/output' );
            const reportFolder = path.join( process.cwd(),'./.tmp/report' );
            const jsonFile = `${outputFolder}/this-feature.json`;

            removeSync( outputFolder );
            expect( fileExists( outputFolder ) ).toEqual( false );
            copySync( 'lib/tests/__mocks__/mock.json', jsonFile );
            tmpReporter = new CucumberHtmlJsonReporter( <Models.ReportGeneration>{ jsonDir: outputFolder, reportPath: reportFolder }, 'en' );
            tmpReporter.report.feature = { id: 'this-feature' };

            await tmpReporter.onRunnerEnd();

            const files = readdirSync( outputFolder );

            expect( files.length ).toEqual( 1 );
            expect( files[0].includes( `${tmpReporter.report.feature.id}.json` ) ).toEqual( true );
            expect( fileExists( jsonFile ) ).toEqual( true );

            // Clean up
            removeSync( outputFolder );
            removeSync( reportFolder );
        } );

        it( 'should be able to add json to an existing json output', async () => {
            const outputFolder = path.join( process.cwd(),'./.tmp/output' );
            const jsonFile = `${outputFolder}/mock.json`;
            removeSync( outputFolder );
            copySync( 'lib/tests/__mocks__/mock.json', jsonFile );

            tmpReporter = new CucumberHtmlJsonReporter( <Models.ReportGeneration>{ jsonDir: outputFolder, reportPath: './.tmp' }, 'en' );
            tmpReporter.report.feature = { id: 'this-feature' };

            expect( ( readJsonSync( jsonFile ) as any[] ).length ).toEqual( 1 );

            await tmpReporter.onRunnerEnd();

            const files = readdirSync( outputFolder );

            expect( files.length ).toEqual( 2 );
            expect( ( readJsonSync( jsonFile ) as any[] ).length ).toEqual( 1 );

            // Clean up
            removeSync( outputFolder );
        } );
    } );

    describe( 'attach', () => {
        let mockStdout: jest.SpyInstance;
        beforeAll( () => {
            process.emit = jest.fn();
            mockStdout = jest.spyOn( process, 'emit' ).mockImplementation();
        } );

        afterEach( () => {
        //   process.emit.mockClear();
            mockStdout.mockClear();
        } );

        it( 'should be able to attach default data', () => {
            CucumberHtmlJsonReporter.attach( 'foo' );

            expect( mockStdout ).toHaveBeenCalledTimes( 1 );
            expect( mockStdout ).toHaveBeenCalledWith( 'wdioCucumberHtmlReporter:attachment', {
                data: 'foo',
                type: TEXT_PLAIN
            } );
        } );

        it( 'should be able to attach with all data', () => {
            CucumberHtmlJsonReporter.attach( 'foo', 'type/string' );

            expect( mockStdout ).toHaveBeenCalledTimes( 1 );
            expect( mockStdout ).toHaveBeenCalledWith( 'wdioCucumberHtmlReporter:attachment', {
                data: 'foo',
                type: 'type/string'
            } );
        } );
    } );

    describe( 'cucumberJsAttachment', () => {
        it( 'should be able to add embeddings to a current step when they have not been added', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            keyword: 'Given',
                            result: {
                                status: PENDING
                            }
                        }]
                    }]
                }
            };

            tmpReporter.cucumberJsAttachment( { data: 'foo', type: 'type/string' } );

            expect( tmpReporter.report.feature.elements[0].steps[0] ).toMatchSnapshot();
        } );

        it( 'should be able to add embeddings to a current step which already has embeddings', () => {
            tmpReporter.report = <Report>{
                feature: {
                    elements: <Scenario[]>[{
                        steps: [{
                            keyword: 'Given',
                            arguments: [{
                                'rows': [
                                    {
                                        'cells': [
                                            'Cucumber',
                                            'Cucumis sativus'
                                        ]
                                    },
                                    {
                                        'cells': [
                                            'Burr Gherkin',
                                            'Cucumis anguria'
                                        ]
                                    }
                                ]
                            }],
                        }]
                    }]
                }
            };

            tmpReporter.cucumberJsAttachment( { data: 'data-2', type: 'mime_type-2' } );
            expect( tmpReporter.report.feature.elements[0].steps[0] ).toMatchSnapshot();
            tmpReporter.cucumberJsAttachment( { data: 'data-2', type: 'mime_type-2' } );

            expect( tmpReporter.report.feature.elements[0].steps[0] ).toMatchSnapshot();
        } );
    } );
} );
