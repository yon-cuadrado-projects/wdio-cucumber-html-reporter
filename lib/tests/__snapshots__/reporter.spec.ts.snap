// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reporter cucumberJsAttachment should be able to add embeddings to a current step when they have not been added 1`] = `
{
  "embeddings": [
    {
      "data": "foo",
      "mime_type": "type/string",
    },
  ],
  "keyword": "Given",
  "result": {
    "status": "pending",
  },
}
`;

exports[`reporter cucumberJsAttachment should be able to add embeddings to a current step which already has embeddings 1`] = `
{
  "arguments": [
    {
      "rows": [
        {
          "cells": [
            "Cucumber",
            "Cucumis sativus",
          ],
        },
        {
          "cells": [
            "Burr Gherkin",
            "Cucumis anguria",
          ],
        },
      ],
    },
  ],
  "embeddings": [
    {
      "data": "data-2",
      "mime_type": "mime_type-2",
    },
  ],
  "keyword": "Given",
}
`;

exports[`reporter cucumberJsAttachment should be able to add embeddings to a current step which already has embeddings 2`] = `
{
  "arguments": [
    {
      "rows": [
        {
          "cells": [
            "Cucumber",
            "Cucumis sativus",
          ],
        },
        {
          "cells": [
            "Burr Gherkin",
            "Cucumis anguria",
          ],
        },
      ],
    },
  ],
  "embeddings": [
    {
      "data": "data-2",
      "mime_type": "mime_type-2",
    },
    {
      "data": "data-2",
      "mime_type": "mime_type-2",
    },
  ],
  "keyword": "Given",
}
`;

exports[`reporter on create should set the defaults only if the reportProperties option is provided 1`] = `
{
  "jsonFolder": "./tmp/jsons",
  "language": "en",
  "logFile": "tmp1/logFile.json",
  "outputDir": "tmp1",
}
`;

exports[`reporter on create should verify initial properties 1`] = `
{
  "jsonFolder": "./tmp/jsons",
  "language": "en",
  "logFile": "./.tmp/logFile.json",
  "outputDir": "./.tmp",
}
`;

exports[`reporter on create should verify initial properties 2`] = `{}`;

exports[`reporter on create should verify initial properties 3`] = `
{
  "feature": {
    "elements": [],
  },
}
`;

exports[`reporter onHookStart should add a pending after step 1`] = `
{
  "arguments": [],
  "keyword": "After",
  "line": "",
  "match": {
    "location": "can not be determined with webdriver.io",
  },
  "name": "",
  "result": {
    "duration": 0,
    "status": "passed",
  },
}
`;

exports[`reporter onHookStart should add a pending before step 1`] = `
{
  "arguments": [],
  "keyword": "Before",
  "line": "",
  "match": {
    "location": "can not be determined with webdriver.io",
  },
  "name": "",
  "result": {
    "duration": 0,
    "status": "passed",
  },
}
`;

exports[`reporter onSuiteStart should add a scenario to the feature if the feature is already there 1`] = `
[
  {
    "description": "",
    "id": "test;test",
    "keyword": "Scenario",
    "name": "test",
    "steps": [],
    "tags": [],
    "type": undefined,
  },
]
`;

exports[`reporter onSuiteStart should add the CucumberJS feature object if it is not available 1`] = `{}`;

exports[`reporter onSuiteStart should add the CucumberJS feature object if it is not available 2`] = `
{
  "feature": {
    "metadata": null,
  },
}
`;
